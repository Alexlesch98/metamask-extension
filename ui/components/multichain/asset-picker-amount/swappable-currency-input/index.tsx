import React from 'react';
import { useDispatch } from 'react-redux';
import { toHex } from '@metamask/controller-utils';
import { useI18nContext } from '../../../../hooks/useI18nContext';
import useIsFiatPrimary from '../hooks/useIsFiatPrimary';
import type { Asset, Amount } from '../../../../ducks/send';
import { toggleCurrencySwitch } from '../../../../ducks/app/app';
import { LARGE_SYMBOL_LENGTH } from '../constants';
import { AssetType } from '../../../../../shared/constants/transaction';
import { Box, Text } from '../../../component-library';
import {
  FontWeight,
  TextVariant,
} from '../../../../helpers/constants/design-system';
import CurrencyInput from '../../../app/currency-input';
import SwapIcon from './swap-icon';

interface BaseProps {
  assetType: AssetType;
  onAmountChange: (newAmount: string) => void;
  asset?: Asset;
  amount: Amount;
}

interface ERC20Props extends BaseProps {
  assetType: AssetType.token;
  asset: Asset;
}

// TODO: build out NFT logic
interface NFTProps extends BaseProps {
  assetType: AssetType.NFT;
  asset: Asset;
}

interface NativeTokenProps extends BaseProps {
  assetType: AssetType.native;
}

interface UnsupportedTokenProps extends Omit<BaseProps, 'assetType'> {
  assetType: Exclude<
    AssetType,
    AssetType.NFT | AssetType.native | AssetType.token
  >;
  asset?: Asset;
}

type SwappableCurrencyInputProps =
  | ERC20Props
  | NFTProps
  | NativeTokenProps
  | UnsupportedTokenProps;

export default function SwappableCurrencyInput({
  assetType,
  asset,
  amount: { value },
  onAmountChange,
}: SwappableCurrencyInputProps) {
  const dispatch = useDispatch();

  const t = useI18nContext();

  const isFiatPrimary = useIsFiatPrimary();

  // FIXME: update swapping logic
  // TODO: add NFTs
  switch (assetType) {
    case AssetType.token:
    case AssetType.native:
      return (
        <CurrencyInput
          className="asset-picker-amount__input"
          featureSecondary={isFiatPrimary}
          onChange={onAmountChange}
          hexValue={toHex(value)}
          swapIcon={(onClick: React.MouseEventHandler) => (
            <SwapIcon ariaLabel={t('switchInputCurrency')} onClick={onClick} />
          )}
          onPreferenceToggle={() => dispatch(toggleCurrencySwitch())}
          hideSuffix={
            !isFiatPrimary ||
            (asset?.details?.symbol?.length || 0) > LARGE_SYMBOL_LENGTH
          }
        />
      );

    default:
      return (
        <Box marginLeft={'auto'}>
          <Text variant={TextVariant.bodySm}>{t('tokenId')}</Text>
          <Text
            variant={TextVariant.bodySm}
            fontWeight={FontWeight.Bold}
            marginLeft={10}
          >
            {asset?.details?.tokenId}
          </Text>
        </Box>
      );
  }
}
