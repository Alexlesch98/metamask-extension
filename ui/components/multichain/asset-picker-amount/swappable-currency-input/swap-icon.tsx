import React from 'react';
import {
  BackgroundColor,
  IconColor,
} from '../../../../helpers/constants/design-system';
import {
  ButtonIcon,
  ButtonIconSize,
  IconName,
} from '../../../component-library';

export default function SwapIcon({
  onClick,
  ariaLabel,
}: {
  ariaLabel: string;
  onClick?: React.MouseEventHandler;
}) {
  return (
    <ButtonIcon
      backgroundColor={BackgroundColor.transparent}
      iconName={IconName.Refresh}
      size={ButtonIconSize.Sm}
      color={IconColor.primaryDefault}
      onClick={onClick}
      ariaLabel={ariaLabel}
    />
  );
}
