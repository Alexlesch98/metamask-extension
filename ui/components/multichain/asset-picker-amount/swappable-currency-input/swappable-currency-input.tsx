import React from 'react';
import { useDispatch } from 'react-redux';
import { useI18nContext } from '../../../../hooks/useI18nContext';
import useIsFiatPrimary from '../hooks/useIsFiatPrimary';
import type { Asset, Amount } from '../../../../ducks/send';
import { toggleCurrencySwitch } from '../../../../ducks/app/app';
import { AssetType } from '../../../../../shared/constants/transaction';
import { Box, Text } from '../../../component-library';
import {
  FontWeight,
  TextVariant,
} from '../../../../helpers/constants/design-system';
import CurrencyInput from '../../../app/currency-input';
import SwapIcon from './swap-icon';

interface BaseProps {
  assetType: AssetType;
  onAmountChange: (newAmount: string) => void;
  asset?: Asset;
  amount: Amount;
}

interface ERC20Props extends BaseProps {
  assetType: AssetType.token;
  asset: Asset;
}

interface NFTProps extends BaseProps {
  assetType: AssetType.NFT;
  asset: Asset;
}

interface NativeTokenProps extends BaseProps {
  assetType: AssetType.native;
}

interface UnsupportedTokenProps extends Omit<BaseProps, 'assetType'> {
  assetType: Exclude<
    AssetType,
    AssetType.NFT | AssetType.native | AssetType.token
  >;
  asset?: Asset;
}

type SwappableCurrencyInputProps =
  | ERC20Props
  | NFTProps
  | NativeTokenProps
  | UnsupportedTokenProps;

export function SwappableCurrencyInput({
  assetType,
  asset,
  amount: { value },
  onAmountChange,
}: SwappableCurrencyInputProps) {
  const dispatch = useDispatch();

  const t = useI18nContext();

  const isFiatPrimary = useIsFiatPrimary();

  switch (assetType) {
    case AssetType.token:
    case AssetType.native:
      return (
        <CurrencyInput
          className="asset-picker-amount__input"
          featureSecondary={isFiatPrimary}
          onChange={onAmountChange}
          hexValue={value}
          swapIcon={(onClick: React.MouseEventHandler) => (
            <SwapIcon onClick={onClick} />
          )}
          onPreferenceToggle={() => dispatch(toggleCurrencySwitch())}
          asset={asset?.details}
        />
      );

    default:
      return (
        <Box marginLeft={'auto'}>
          <Text variant={TextVariant.bodySm}>{t('tokenId')}</Text>
          <Text
            variant={TextVariant.bodySm}
            fontWeight={FontWeight.Bold}
            marginLeft={10}
          >
            {asset?.details?.tokenId}
          </Text>
        </Box>
      );
  }
}
