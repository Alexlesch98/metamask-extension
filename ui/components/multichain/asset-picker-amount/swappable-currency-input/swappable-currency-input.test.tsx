import React from 'react';
import { getByRole, render } from '@testing-library/react';
import { SwappableCurrencyInput } from './swappable-currency-input';
import { AssetType } from '../../../../../shared/constants/transaction';
import mockSendState from '../../../../../test/data/mock-send-state.json';
import { Provider } from 'react-redux';
import configureStore from '../../../../store/store';

const createStore = ({
  useNativeCurrencyAsPrimaryCurrency,
  sendInputCurrencySwitched,
}) =>
  configureStore({
    ...mockSendState,
    metamask: {
      ...mockSendState.metamask,
      preferences: { useNativeCurrencyAsPrimaryCurrency },
    },
    appState: { ...mockSendState.appState, sendInputCurrencySwitched },
  });

describe('SwappableCurrencyInput', () => {
  it('native: matches snapshot', () => {
    const asset = {
      type: AssetType.native,
      balance: '1000000',
    };
    const mockAssetChange = jest.fn();

    const { asFragment, getByText } = render(
      <Provider
        store={createStore({
          useNativeCurrencyAsPrimaryCurrency: true,
          sendInputCurrencySwitched: true,
        })}
      >
        <SwappableCurrencyInput
          asset={asset}
          assetType={AssetType.native}
          amount={{ value: '5000' }}
          onAmountChange={mockAssetChange}
        />
      </Provider>,
    );
    expect(asFragment()).toMatchSnapshot('native');
    expect(getByText('ETH')).toBeInTheDocument();
  });

  it('token: matches snapshot', () => {
    const asset = {
      type: AssetType.token,
      details: {
        address: '0x0bc529c00c6401aef6d220be8c6ea1667f6ad93e',
        symbol: 'TOKEN',
        isERC721: false,
      },
      balance: '1000000',
    };
    const mockAssetChange = jest.fn();

    const { asFragment, getByText } = render(
      <Provider
        store={createStore({
          useNativeCurrencyAsPrimaryCurrency: true,
          sendInputCurrencySwitched: true,
        })}
      >
        <SwappableCurrencyInput
          asset={asset}
          assetType={AssetType.native}
          amount={{ value: '5000' }}
          onAmountChange={mockAssetChange}
        />
      </Provider>,
    );
    expect(asFragment()).toMatchSnapshot('token');
    expect(getByText('TOKEN')).toBeInTheDocument();
  });

  it('ERC721: matches snapshot', () => {
    const asset = {
      type: AssetType.token,
      details: {
        address: '0x0bc529c00c6401aef6d220be8c6ea1667f6ad93e',
        symbol: 'TOKEN',
        isERC721: true,
        tokenId: 1234,
        decimals: 0,
      },
      balance: '1000000',
    };
    const mockAssetChange = jest.fn();

    const { asFragment, getByText } = render(
      <Provider
        store={createStore({
          useNativeCurrencyAsPrimaryCurrency: true,
          sendInputCurrencySwitched: true,
        })}
      >
        <SwappableCurrencyInput
          asset={asset}
          assetType={AssetType.NFT}
          amount={{ value: '5000' }}
          onAmountChange={mockAssetChange}
        />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot('nft');
    expect(getByText('1234')).toBeInTheDocument();
  });
});
