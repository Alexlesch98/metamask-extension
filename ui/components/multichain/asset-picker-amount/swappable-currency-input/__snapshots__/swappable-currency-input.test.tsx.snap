// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwappableCurrencyInput ERC721: matches snapshot: nft 1`] = `
<DocumentFragment>
  <div
    class="mm-box mm-box--margin-left-auto"
  >
    <p
      class="mm-box mm-text mm-text--body-sm mm-box--color-text-default"
    >
      [tokenId]
    </p>
    <p
      class="mm-box mm-text mm-text--body-sm mm-text--font-weight-bold mm-box--margin-left-10 mm-box--color-text-default"
    >
      1234
    </p>
  </div>
</DocumentFragment>
`;

exports[`SwappableCurrencyInput native: matches snapshot: native 1`] = `
<DocumentFragment>
  <div
    class="unit-input asset-picker-amount__input"
  >
    <div
      class="unit-input__inputs"
    >
      <div
        class="unit-input__input-container"
        style="display: inherit;"
        title="0.00000000000002048"
      >
        <input
          class="unit-input__input"
          data-testid="currency-input"
          dir="ltr"
          placeholder="0"
          style="width: 19ch;"
          type="number"
          value="0"
        />
        <div
          class="unit-input__suffix"
        >
          ETH
        </div>
      </div>
      <div
        class="currency-input__conversion-component"
      >
        [noConversionRateAvailable]
      </div>
    </div>
    <button
      aria-label="[switchInputCurrency]"
      class="mm-box mm-button-icon mm-button-icon--size-sm mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-primary-default mm-box--background-color-transparent mm-box--rounded-lg"
    >
      <span
        class="mm-box mm-icon mm-icon--size-sm mm-box--display-inline-block mm-box--color-inherit"
        style="mask-image: url('./images/icons/swap-vertical.svg');"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`SwappableCurrencyInput token: matches snapshot: token 1`] = `
<DocumentFragment>
  <div
    class="unit-input asset-picker-amount__input"
  >
    <div
      class="unit-input__inputs"
    >
      <div
        class="unit-input__input-container"
        style="display: inherit;"
        title="0.00000000000002048"
      >
        <input
          class="unit-input__input"
          data-testid="currency-input"
          dir="ltr"
          placeholder="0"
          style="width: 19ch;"
          type="number"
          value="0"
        />
        <div
          class="unit-input__suffix"
        >
          TOKEN
        </div>
      </div>
      <div
        class="currency-input__conversion-component"
      >
        [noConversionRateAvailable]
      </div>
    </div>
    <button
      aria-label="[switchInputCurrency]"
      class="mm-box mm-button-icon mm-button-icon--size-sm mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-primary-default mm-box--background-color-transparent mm-box--rounded-lg"
    >
      <span
        class="mm-box mm-icon mm-icon--size-sm mm-box--display-inline-block mm-box--color-inherit"
        style="mask-image: url('./images/icons/swap-vertical.svg');"
      />
    </button>
  </div>
</DocumentFragment>
`;
